src/protocol/sort/bit_permutations.rs

```diff
@@ -0,0 +1,156 @@
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::{context::ProtocolContext, IPAProtocolStep, RecordId, SortStep},
+    secret_sharing::Replicated,
+};
+
+use futures::future::try_join_all;
+
+#[derive(Debug)]
+pub struct BitPermutations<'a, F> {
+    input: &'a [Replicated<F>],
+}
+
+impl<'a, F: Field> BitPermutations<'a, F> {
+    /// Create an object to generate bit permutations for a given bit column of query. This is GENBITPERM(Algorithm 3) from the paper
+    #[allow(dead_code)]
+    pub fn new(input: &'a [Replicated<F>]) -> BitPermutations<'a, F> {
+        Self { input }
+    }
+
+    /// In this step, multiplication inputs are generated locally at each helper from input share, x in following steps
+    /// 1. calculate 1 - x, x and concatenate them
+    /// 2. calculate cumulative sum at each vector row
+    /// 3. return back tuple of step 1 and step 2 output
+    fn prepare_mult_inputs<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
+        &self,
+        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
+    ) -> Vec<(Replicated<F>, Replicated<F>)>
+    where
+        F: Field,
+    {
+        let channel = ctx
+            .gateway
+            .get_channel(IPAProtocolStep::Sort(SortStep::BitPermutations));
+
+        self.input
+            .iter()
+            .map(|x| channel.share_of_one() - *x)
+            .chain(self.input.iter().copied())
+            .scan(Replicated::<F>::new(F::ZERO, F::ZERO), |sum, n| {
+                *sum = *sum + n;
+                Some((n, *sum))
+            })
+            .collect()
+    }
+
+    /// multiplies the input vector pairs across helpers and returns result
+    /// For this, it spawns all multiplication, wait for them to finish in parallel and then collect the results
+
+    #[allow(clippy::cast_possible_truncation)]
+    async fn secure_multiply<M: Mesh, G: Gateway<M, IPAProtocolStep>>(
+        &self,
+        ctx: &ProtocolContext<'_, G, IPAProtocolStep>,
+        mult_inputs: Vec<(Replicated<F>, Replicated<F>)>,
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        F: Field,
+    {
+        let async_multiply = mult_inputs
+            .iter()
+            .enumerate()
+            .map(|(index, input)| async move {
+                ctx.multiply(
+                    RecordId::from(index as u32),
```

Comment:
I think record id should be provided from outside, i.e. our input is `(RecordId, Replicated<F>)`